Features!

justynr opened the pull request:
Added a few new features and modified some existing ones.
It has all tested and working in the latest MM software.
Removed a call that caused a double reload of the Dom. (commented out the line and added explanation)

### Added:

#### Field
- `map` - field - Google Map Display - Shows a static map (a picture) with a pin signifying where the vehicle is.
  - If the vehicle is parked, the map will be translucent (dimmed)
  - `excludeLocations` - config option - Users can hide the map when the vehicle is at a tagged location - enter tags here - case-insensitive
  - `googleMapApiKey` - config option
  - `mapZoom` - config option - set zoom of map
  - `mapWidth` & `mapHeight` - config option - set size of map in pixels

- `state` - shows the state of the vehicle (Idling/Driving/Sentry)
  - when 'Driving' shows the vehicle's speed and heading (North/North East/...)

- `version` - shows the currently installed version of software
  - when a new version is available to download or ready to install, this field turns green and says "New Software" and its version

- `newVersion` - only displayed (in green) if a new version is available for download or waiting to install. 

- `location` - displays the location tag from TeslaFi: 
  - if parked
  - if location is tagged 
  - if the car is parked at an untagged location, this field is hidden.
  - if the car is driving this field is hidden (TeslaFi reports it's at the last position until it parks at a new one)

- `charge-power` - displays power level (in kW) if charging

- `distance` - displays vehicle's distance from "homeAddress" (as route, not straight line)
- `duration` - displays duration (in traffic) it would take to drive from vehicle's current location to "homeAddress"

  - if distance, duration, or map are not used, no Google API Key is needed


#### Config Options
##### Changed:
- `lock` - now hidden if vehicle is driving (it's always locked when driving)

##### Added:

- `dataTimeout` - hides `data-time` if the data is younger than the time set
How old data must be in seconds before 'data-time' is displayed (in seconds).
  - default is to always show this if this value is not set.
  - (found that this usually showed 'a few seconds ago' - personally I only want to know if data is >5 minutes old)


- `googleMapApiKey` - used for map, distance and duration 
  - must activate "Maps Static API" for the map & "Distance Matrix API" for the distance and duration fields

- `mapZoom` - Sets the zoom level of map (default = 13)

- `mapWidth` - Sets the width of map in pixels (default = 300)
- `mapHeight` Sets the height of map in pixels (default = 150)

- `excludeLocations` - ["array", "of", "locations", "to", "exclude"] - any locations set in here will stop the plugin from using any Google APIs - this will also hide the 'map', 'distance' and 'duration' fields. (case insensitive)

- `homeAddress` - home address of mirror - used for 'distance' and 'duration' calculations - must use an address that can be found on Google Maps. ie. `123 Fake St. Toronto ON Canada`
  - if blank, 'distance' and 'duration' will be skipped

mattdy reviewed (changes requested):
This commit seems to have changed a lot of the spacing in `MMM-TeslaFi.js` - I appreciate that some code clean-up may be required, but I'd prefer to do that separately rather than overwriting everything like this. Any chance you could revert this and only include code changes?

Thanks

mattdy commented on MMM-TeslaFi.js:
```diff
@@ -281,5 +307,25 @@ Module.register('MMM-TeslaFi', {
+	// Converts given speed (assumes miles input) to configured output with approprate units appened
+	convertSpeed: function(valueMiles) {
+		if(this.config.unitDistance==="km") {
+			return this.numberFormat(valueMiles * 1.60934) + " Km/h"
```
Could you change this to the correct SI unit of "km/h" rather than "Km/h"?

justynr replied:
done, good catch

mattdy commented on MMM-TeslaFi.js:
```diff
@@ -281,5 +307,25 @@ Module.register('MMM-TeslaFi', {
+		if(this.config.unitDistance==="km") {
+			return this.numberFormat(valueMiles * 1.60934) + " Km/h"
+		} else {
+			return this.numberFormat(valueMiles) + " Mph"
```
As above, could you please change this to "mph"?

justynr replied:
Just pushed this. done.
Good catch!

mattdy reviewed (changes requested):
Looks great, could you just update the units to match the rest of the module?

mattdy commented on README.md:
```diff
@@ -11,6 +11,9 @@ I am happy to accept any [bug reports](https://github.com/mattdy/MMM-TeslaFi/iss
 ## Screenshot
 
 ![screenshot.png](doc/screenshot.png)
+![screenshot.png](doc/Driving-With-Map.png)
```
Could this be improved by adding an caption saying what options are selected on each screenshot? Looks a little confusing otherwise!

justynr replied:
I've added which fields are active in each screenshot (updated one and removed a redundant screenshot)

mattdy commented:
Awesome new features, thank you so much! I've just requested a couple of changes if you wouldn't mind? Then I'd be delighted to merge this in!

justynr commented:
I didn't overwrite anything here really, I added each feature in its own commit (I've not really made such a large pull request like this before)
Would it be better to have one pull request for each feature next time?

I added each field to the bottom of your loop/switch: case part (clever solution btw), added the necessary functions to the very bottom of the document, and followed the same naming conventions.

One thing that I think did change a lot in here - MitchMitch's Prettier that was added to the whole MagicMirror project.
All of my quotes were single; I just noticed they're not anymore... more was likely changed.

The big thing that I found quite messy was the Google Matrix API in the Node_Helper.js file.
I couldn't help it as I've got to parse the JSON there, get Long/Lat. Send the request to Google, parse the response, push the object to the TeslaFi Object, then stringify it before sending it off... haha

I'm open to suggestions on what you'd like me to do to this file. (email works as well)

mattdy commented:
> One thing that I think did change a lot in here - MitchMitch's Prettier that was added to the whole MagicMirror project.

I think this is probably the cause of the large diff - looks like you ran it then committed in 96d7392be7a06ab0c3e1812ab44335759cc017cc. Would you mind backing that change out, and re-committing just the code changes rather than including the reformatting? I'll then look at running `Prettier` at a later date to clean things up.

> Would it be better to have one pull request for each feature next time?

It would make the review slightly easier in future, but I definitely don't mind picking through :-) 

> The big thing that I found quite messy was the Google Matrix API in the Node_Helper.js file.

No worries, I'll take another look at this and see if there's anything I can think of!

Thanks once again!

mattdy commented:
I've run `Prettier` across the files, unfortunately that's caused some merge conflicts on this PR!

Are you able to merge the latest changes back in to your repo, then I should be able to apply this PR?

Many thanks!

mattdy commented:
After some more discussion, we've agreed to close this PR and submit separate PRs for individual features :-)
